Haskell Parser Combinators/article.md

```diff
@@ -0,0 +1,1330 @@
+
+Begin by defining `skipSpace`. We use the `space` helper function from the lexer module, which expects a function to parse 1 or more spaces, a line comment, and a block comment.
+
+```hs
+skipSpace :: Parser ()
+skipSpace = L.space
+  -- Like `space`, but skips 1 or more space characters.
+  space1
+  -- Skip from ;; until a newline.
+  (L.skipLineComment ";;")
+  -- Skip from /* until */. There is also `skipBlockComment`, but it doesn't handle nested comments.
+  (L.skipBlockCommentNested "/*" "*/")
+```
+
+And now define a `lexeme` parser, which indicates how to consume space after the given parser.
+
+```hs
+lexeme :: Parser a -> Parser a
+lexeme = L.lexeme skipSpace
+```
+
+Now you can use `lexeme` in each of the parsers, and they will automatically skip whitespace after their definitions. Note we also use `lexeme` before our first opening parenthesis, so we can skip space just after the `(`.
+
+```hs
+sexp :: Parser (SExp, [SExp])
+sexp = label "S-expression" $ lexeme $
+  between (lexeme (char '(')) (char ')') ((,) <$> atom <*> many atom)
+```
+
+You may add `lexeme` in other parsers as well, where it's desirable to skip whitespace after that rule. For instance:
+
+```hs
+integer :: Parser Integer
+integer = label "integer" $ lexeme $ read <$> some numberChar
+```
+
+**Note**: The reader should do the same for `str`, `bool`, and `identifier`.
+
+And now, we can parse S-expressions properly:
+
+```hs
+>>> parseTest atom "(  foo  /* 8-) */  bar  ;; look mom, comments! \n  )"
+SSExp (SId (Identifier {getId = "foo"})) [SId (Identifier {getId = "bar"})]
+```
+
+### Our main parser function
+
+Finally, let's make a function that solves the last two outstanding problems.
+
+For the first one, as mentioned by the Megaparsed documentation, `lexeme` will skip whitespace after a parser, but we also need to deal with whitespace at the beginning of the file.
+
+For the second one, we want to parse everything until we find an `eof` since our parser will currently ignore anything after a complete atom. The two defects are exemplified below:
+
+```hs
+>>> parseTest atom "  42"
+1:1:
+  |
+1 |   42
+  | ^^^^
+unexpected "  42"
+expecting S-expression, boolean, identifier, integer, or string
+
+>>> parseTest atom "42 foo bar"
+SInteger 42
+```
+
+We will create a function that takes our input string, and either returns an error message or the actual result. Such function "bootstraps" our parser with `between skipSpace eof atom`, which will solve the two defects. It will also serve as an auxiliary function that can be called by client code, rather than `parseTest` which is useful for working on GHCi.
```

Comment:
```suggestion
We will create a function that takes our input string, and either returns an error message or the actual result. Such function "bootstraps" our parser with `between skipSpace eof atom`, which will solve the two defects. It will also serve as an auxiliary function that can be called by client code, rather than `parseTest` which is more useful for working in GHCi.
```

Or some other changes that show contrast, otherwise just saying that it is useful for ghci sounds a bit odd.